File: backend/services/admin_content.go
package services

import (
	"strings"

	"github.com/affiliatedonor/backend/models"
)

// SummarizeAdminContent returns status rollups for CMS blocks.
func SummarizeAdminContent(blocks []models.AdminContentBlock) models.AdminContentSummary {
	summary := models.AdminContentSummary{}
	for _, block := range blocks {
		summary.TotalCount++
		switch block.Status {
		case "published":
			summary.PublishedCount++
		case "draft":
			summary.DraftCount++
		case "scheduled":
			summary.ScheduledCount++
		case "archived":
			summary.ArchivedCount++
		}
		if contentBlockNeedsMedia(block.Type) && block.MediaURL == "" {
			summary.MissingMediaCount++
		}
	}
	return summary
}

// FilterAdminContent filters blocks by common admin query parameters.
func FilterAdminContent(blocks []models.AdminContentBlock, area string, status string, blockType string, search string) []models.AdminContentBlock {
	filtered := make([]models.AdminContentBlock, 0, len(blocks))
	normalizedSearch := strings.ToLower(strings.TrimSpace(search))

	for _, block := range blocks {
		if area != "" && block.Area != area {
			continue
		}
		if status != "" && status != "all" && block.Status != status {
			continue
		}
		if blockType != "" && blockType != "all" && block.Type != blockType {
			continue
		}
		if normalizedSearch != "" && !contentBlockMatchesSearch(block, normalizedSearch) {
			continue
		}
		filtered = append(filtered, block)
	}

	return filtered
}

func contentBlockMatchesSearch(block models.AdminContentBlock, search string) bool {
	searchable := strings.ToLower(strings.Join([]string{
		block.Title,
		block.Slug,
		block.Area,
		block.Type,
		block.Status,
		block.Summary,
		block.Body,
		block.LinkedEntityLabel,
	}, " "))
	return strings.Contains(searchable, search)
}

func contentBlockNeedsMedia(blockType string) bool {
	return blockType == "homepage_hero" || blockType == "banner" || blockType == "impact_story" || blockType == "testimonial"
}

File: src/types/adminContent.ts
export type AdminContentStatus = 'draft' | 'published' | 'scheduled' | 'archived'

export type AdminContentArea =
  | 'homepage'
  | 'banners'
  | 'impact-stories'
  | 'testimonials'
  | 'about'
  | 'footer'

export type AdminContentType =
  | 'homepage_hero'
  | 'announcement'
  | 'banner'
  | 'impact_story'
  | 'testimonial'
  | 'about_section'
  | 'footer_group'
  | 'featured_section'

export type AdminContentSort = 'newest' | 'oldest' | 'title' | 'status' | 'order'

export type AdminContentMetadataValue = string | number | boolean | string[] | null

export type AdminContentMetadata = Record<string, AdminContentMetadataValue>

export interface AdminContentBlock {
  id: string
  area: AdminContentArea
  type: AdminContentType
  title: string
  slug: string
  status: AdminContentStatus
  summary: string
  body: string
  mediaUrl?: string
  ctaLabel?: string
  ctaTarget?: string
  linkLabel?: string
  linkTarget?: string
  linkedEntityId?: string
  linkedEntityLabel?: string
  metadata: AdminContentMetadata
  sortOrder: number
  startAt?: string
  endAt?: string
  scheduledAt?: string
  publishedAt?: string
  archivedAt?: string
  updatedBy: string
  createdAt: string
  updatedAt: string
}

export interface AdminContentSummary {
  totalCount: number
  publishedCount: number
  draftCount: number
  scheduledCount: number
  archivedCount: number
  missingMediaCount: number
}

export interface AdminContentListResponse {
  blocks: AdminContentBlock[]
  summary: AdminContentSummary
}

export interface AdminContentFilters {
  search: string
  area: AdminContentArea | 'all'
  type: AdminContentType | 'all'
  status: AdminContentStatus | 'all'
  scheduled: 'all' | 'scheduled' | 'unscheduled'
  updatedFrom: string
  updatedTo: string
  sort: AdminContentSort
}

export interface AdminContentFilterOptions {
  areas: AdminContentArea[]
  types: AdminContentType[]
  statuses: AdminContentStatus[]
}

export interface AdminContentAreaConfig {
  area: AdminContentArea
  title: string
  description: string
  path: string
  allowedTypes: AdminContentType[]
}

export interface AdminContentEditorProps {
  draft: AdminContentBlock
  onFieldChange: <Key extends keyof AdminContentBlock>(key: Key, value: AdminContentBlock[Key]) => void
  onMetadataChange: (key: string, value: AdminContentMetadataValue) => void
}
